spring-web/src/main/java/org/springframework/http/server/reactive/AbstractListenerReadPublisher.java

```diff
@@ -256,6 +262,7 @@ private State state() {
 						}
 						else {
 							publisher.changeState(READING, NO_DEMAND);
+							publisher.suspendReading();
```

Comment:
@violetagg if we switch to NO_DEMAND first and then suspend, isn't it possible that if a request comes in concurrently, after NO_DEMAND but before suspend, it will checkOnDataAvailable, but then the suspend after may end up cancelling the request for data? It seems safer to suspend first and then change the state.

if we switch to NO_DEMAND first and then suspend reading, isn't there a possibility that in between a request may come which would call check

Reply:
Also shouldn't suspend be called in UNSUBSCRIBED too before there is demand? It should apply every time we're about to switch to NO_DEMAND.